src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,860 @@
+        let config = MongodbConfig::from_hashmap(param.properties.clone())?;
+        let pk_indices = param.downstream_pk.clone();
+        let is_append_only = param.sink_type.is_append_only();
+        let schema = param.schema();
+        Ok(Self {
+            config,
+            param,
+            schema,
+            pk_indices,
+            is_append_only,
+        })
+    }
+}
+
+impl TryFrom<SinkParam> for MongodbSink {
+    type Error = SinkError;
+
+    fn try_from(param: SinkParam) -> std::result::Result<Self, Self::Error> {
+        MongodbSink::new(param)
+    }
+}
+
+impl Sink for MongodbSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type LogSinker = LogSinkerOf<MongodbSinkWriter>;
+
+    const SINK_NAME: &'static str = MONGODB_SINK;
+
+    fn is_sink_decouple(desc: &SinkDesc, user_specified: &SinkDecouple) -> Result<bool> {
+        match user_specified {
+            SinkDecouple::Default => Ok(desc.sink_type.is_append_only()),
+            SinkDecouple::Disable => Ok(false),
+            SinkDecouple::Enable => Ok(true),
+        }
+    }
+
+    async fn validate(&self) -> Result<()> {
+        if !self.is_append_only {
+            if self.pk_indices.is_empty() {
+                return Err(SinkError::Config(anyhow!(
+                    "Primary key not defined for upsert mongodb sink (please define in `primary_key` field)")));
+            }
+
+            // checking if there is a non-pk field's name is `_id`
+            if self
+                .schema
+                .fields
+                .iter()
+                .enumerate()
+                .any(|(i, field)| !self.pk_indices.contains(&i) && field.name == MONGODB_PK_NAME)
+            {
+                return Err(SinkError::Config(anyhow!(
+                    "_id field must be the sink's primary key, but a non primary key field name is _id",
+                )));
+            }
+
+            // assume the sink's pk is (a, b) and then the data written to mongodb will be
+            // { "_id": {"a": 1, "b": 2}, "a": 1, "b": 2, ... }
+            // you can see that the compound pk (a, b) is turned into an Object {"a": 1, "b": 2}
+            // and the each pk field is become as a field of the document
+            // but if the sink's pk is (_id, b) and the data will be:
+            // { "_id": {"_id": 1, "b": 2}, "b": 2, ... }
+            // in this case, the original _id field of the compound pk has been overridden
+            // we should consider this is a schema error
+            if self.pk_indices.len() > 1
+                && self
+                    .pk_indices
+                    .iter()
+                    .map(|&idx| self.schema.fields[idx].name.as_str())
+                    .any(|field| field == MONGODB_PK_NAME)
+            {
+                return Err(SinkError::Config(anyhow!(
+                    concat!("the _id field of the sink's primary key will be overridden, please either make the sink's ",
```

Comment:
Didn't get what the overwritten means here. I think the point is that if user specify multiple columns in the `primary_key` field, there should not has a column named `_id`. e.g. `primary_key=_id, col2` is not allowed.

Reply:
You're right, we may not need to hint to the user too specifically just telling the user compound pk's field name should not be `_id` is enough.